packages/api/db/migration/20240524031314_create_animal_identifier_table_and_modify_identifier_placement_table.js

```diff
@@ -0,0 +1,125 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+
+export const up = async function (knex) {
+  await knex.schema.createTable('animal_identifier_type', (table) => {
+    table.increments('id').primary();
+    table.string('key').notNullable();
+  });
+
+  await knex.raw(`
+    CREATE TEMPORARY TABLE temp AS
+    SELECT * FROM animal_identifier_placement;
+  `);
+
+  await knex.schema.alterTable('animal', (t) => {
+    t.dropForeign('identifier_placement_id');
+  });
+
+  await knex.schema.dropTable('animal_identifier_placement');
+
+  await knex.schema.createTable('animal_identifier_placement', (table) => {
+    table.increments('id').primary();
+    table
+      .integer('identifier_type_id')
+      .references('id')
+      .inTable('animal_identifier_type')
+      .notNullable();
+    table.string('key').notNullable();
+  });
+
+  await knex('animal_identifier_type').insert([
+    { id: 1, key: 'EAR_TAG' },
+    { id: 2, key: 'LEG_BAND' },
+    { id: 3, key: 'OTHER' },
+  ]);
+
+  await knex('animal_identifier_placement').insert([
+    { identifier_type_id: 1, key: 'LEFT_EAR' },
+    { identifier_type_id: 1, key: 'RIGHT_EAR' },
+    { identifier_type_id: 2, key: 'LEFT_LEG' },
+    { identifier_type_id: 2, key: 'RIGHT_LEG' },
+    { identifier_type_id: 3, key: 'OTHER' },
+  ]);
```

Comment:
It's funny because it is the opposite of the `permission_id` thing haha. 😢 

@SayakaOno I think you are right for the enum table `animal_identifier_type` we would prefer to use the auto-increment index. And then map that to the correct place in `animal_identifier_placement`. 

If I recall many of our problems with setting this would be migration/rollback/testing related. I think another issue was if it was a user interactable table or not... 

Here is one such initial conversation: https://github.com/LiteFarmOrg/LiteFarm/pull/2857#discussion_r1313469166 and The PR that repaired that id problem: https://github.com/LiteFarmOrg/LiteFarm/pull/2882

I always prefer using the autoincrement but for permission_id both Jimmy and Anto have kept the existing way...

![Screenshot 2024-08-02 at 10 07 48 AM](https://github.com/user-attachments/assets/fcac6284-8650-47ad-b34c-e49a813007cd)

Reply:
I'm still not quite sure I understand why manually setting id in the permissions table is done the way it is. Is it even possible for local/beta/production db to have different id's if using auto-increment? Especially since users cannot edit that table so I'm expecting that it remains stable across environments. I think there are other tables in the schema which aren't able to be edited by a user and yet they still use the auto-increment, so I'm unsure what makes permissions table a special case?
 
Regarding the id's on `animal_identifier_type`, I agree it should've been set with the auto-increment sequence. I think I was just lazy and thought that if its an enum table that won't change then it doesn't matter. But if it caused problems in the past then should probably be changed to auto-increment.

Reply:
So at least for me I have seen auto-increment differences because of failed migrations or undoing-redoing migrations. It seems like because of concurrent transactions the nextval can never be rolled back as part of a transaction. https://www.postgresql.org/docs/9.1/functions-sequence.html

But agreed that we shouldn't be reliant on it.. maybe a SPIKE jira ticket can be made to explore fixing this for good and addressing all concerns.

Reply:
Interesting I didn't know sequence couldn't be rolled back, thanks for sharing the link. In that case it makes sense for manually entering the permission keys, and also for tables like `role` where the id needs to remain stable. But for other tables where we don't care about the value of the id, and only that it's unique in the table, then I think it doesn't really matter if there are differences between environments.